sdk/core/azure-core/src/main/java/com/azure/android/core/util/CancellationToken.java

```diff
@@ -25,15 +31,42 @@ public static CancellationToken create() {
         return new CancellationTokenImpl();
     }
 
+    /**
+     * Creates a new {@link CancellationToken} with a given timeout. 0 (zero) means there is no timeout.
+     *
+     * @return The {@link CancellationToken}.
+     */
+    public static CancellationToken createWithTimeout(@NonNull Duration timeout) {
+        return new CancellationTokenImpl(timeout);
+    }
+
     /**
      * Communicates a request for cancellation.
      */
     public abstract void cancel();
 
+    /**
+     * Start the cancellation token countdown. If the countdown is already running, this should return immediately.
+     */
+    public abstract void start();
+
+    /**
+     * Reset the {@link CancellationToken} and allow the timeout countdown to be restarted.
+     */
+    public abstract void reset();
+
+    /**
+     * Gets this {@link CancellationToken}'s timeout in seconds.
+     */
+    public Duration getTimeout() {
+        return timeout;
+    }
```

Comment:
I see that we want to support a token that triggers a cancellation when timeout happens. It's a "specialized cancellation token" variant of "general cancellation token". 

Why we want to expose the APIs - start|reset -, etc.. specific to the "specialized cancellation token" in the "general cancellation token"? should we not-expose them now and enable it when there is a requirement?

i.e. should we start with a minimal set of public API for the timeout use case? i.e. just 

```java
createWithTimeout(@NonNull Duration timeout);
```

then we hide/remove all other APIs.

thoughts?

Reply:
Another option to consider is to have a setter ".setCancelTimeout(Duration timeout)" in the CancellationToken.

```java
CancellationToken cancelToken = CancellationToken.create();
token.setCancelTimeout(timeout);
apiCall(params, cancelToken);
```

The "setCancelTimeout" can be called multiple times; it will reset the countdown if token not already canceled. This should help us to cut those APIs, at the same time provide the ability to extend the timeout.

Reply:
Also, it is important to note that once a token is canceled, it's canceled forever; we shouldn't implicitly or explicitly allow setting such a token back to be in a non-canceled state.